### ISSUE TITLE
`dd.read_csv` fails when the server doesn't support HTTP HEAD requests

### ISSUE THREAD

**Issue by ian-r-rose** [dask/dask#4633]
Reading a CSV from a URL with `dask.dataframe` fails when the HTTP server does not support HEAD requests. The corresponding `pd.read_csv` succeeds, since it gets the whole file in one go and does not try to chunk the download.

Here is a MWE:
```python
import dask.dataframe as dd
dd.read_csv('https://data.lacity.org/resource/hpbr-2ant.csv', blocksize=None)
```

I had hoped that by setting `blocksize=None` I might be able to download the whole CSV and avoid the HEAD request (GET works fine), but it still tries to read a sample from `b_sample`:
https://github.com/dask/dask/blob/cb58d97f9452f67a922ba332b7a9471289ba94ee/dask/dataframe/io/csv.py#L341-L355
Relevant xrefs: #3160, intake/intake#302

Also, the relevant server (in this case a Socrata data portal) *should* be better about HTTP verbs, but in the absence of that, we can be more fault tolerant.

With a bit of guidance, I'd be happy to submit a PR for this.

**Comment by martindurant** [comment]
The trouble is, that Dask wants to know the metadata of the dataframe before reading all of the data, so that you can do symbolic operations on it in a lazy way. When you load with pandas, you are up-front committed to reading the whole thing.

The version of HTTPFileSystem in [fsspec](https://filesystem-spec.readthedocs.io/en/latest/api.html#fsspec.implementations.http.HTTPFileSystem) is a little smarter than the previous version still used by Dask, in that it will still read the starts of files where less data is fetched than the readahead size. 
However, it still doesn't work, because the "sample" part of read_csv looks for the delimiter around the end of the sample range. This involves seeking to the end of the file, and we don't know where that it. Unfortunately, the server also ignores "range" requests...

In short, the process is geared towards finding line-delimiters in short parts of much bigger files, and then using that information to load the whole thing in relatively large chunks - this is certainly not the case here. At the very best, it would mean downloading all the data twice, once for the metadata and once when actually processing.

**Comment by mrocklin** [comment]
If we were given a list of many (hopefully small) CSV files presumably we
would only have to read in the first file.  Is this correct?

**Comment by ian-r-rose** [comment]
To be sure, if the server doesn't support grabbing header information or range requests, we can't do any of the chunking or lazy operations (at least until it is downloaded).

But is it possible to make this more fault-tolerant for those cases? It would still be nice to be able to read these CSVs, even if the process is not very smart or high-performance.

**Comment by martindurant** [comment]
https://github.com/martindurant/filesystem_spec/pull/38 would help int he fsspec implementation. This could be ported to dask, but is still ugly and hacky. We can be smarter, e.g., have "sample" be the whole of the first file and not try to seek delimiters anywhere. Should be possible with a little work.

**Comment by mrocklin** [comment]
I think that @ian-r-rose is volunteering to do some of that work.  @martindurant if you're able to lay out steps that a newcomer to the project (though generally an experienced dev otherwise) could execute then I suspect that you could get Ian to help you with this.  It may be that this gets too deep into the project though, in which case perhaps it's better not to do so.

(for context, Ian is a JupyterLab dev (among other things) the creator of dask-labextension, and generally a good catch if we can snag his time :) )

**Comment by martindurant** [comment]
I think basically it comes down to being able to pass `sample=False` in [this line](https://github.com/dask/dask/blob/master/dask/dataframe/io/csv.py#L341), in which case the first of the `values` after line 360 would, once computed, play the part of the `b_sample`. Currently, the input `sample` arg to read_csv is treated explicitly as a number further up the function.

**Comment by martindurant** [comment]
Another thing we could consider to shortcut the whole issue, is for the user to be able to provide `meta=` and not bother analysing the header at all.
